### ISSUE TITLE
NetworkManager error when using ExternalCacheProvider

### ISSUE THREAD

**Issue by jaggaer-nkrunic** [coinbase/rest-hooks#638]
Hi,
we've been using rest-hooks v4 for some time now and it's been working quite well for us so far and today we wanted to update to v5 but not everything works after doing the upgrade.
We have followed upgrade guide and migrated everything as described there. However, looks like there are some changes in v5 which are not properly documented.

First we hit the wall with NetworkManager and reducer being moved to @rest-hooks/core, but docs still state to import it from 'rest-hooks'. With that import changed, we faced a second error, coming from 'rest-hooks/lib/manager/index.js'. It tries to import DevToolsConfig from DevToolsManager, but DevToolsConfig is type definition so it doesn't exist after build (we're not using typescript).
Third problem, we are using redux as external cache store and rest-hooks slice is mounted on store sub path. With this config, NetworkManager throws an error because it calls getState() and that returns state from store's root but we in fact need a slice. I went through all the docs and source code but looks like there's currently no way to pass a selector to the NetworkManager.

Have anyone else experienced these issues and if so we'd appreciate any kind of info on how to fix these.

**Comment by ntucker** [comment]
Thanks for the report @jaggaer-nkrunic ! Excited to resolve this and get you on v5 as soon as possible!

1. Updating the docs now, and adding to migration guide
2. Fixed in #619 will release later today
3. Good catch - our tests don't nest the data so we didn't test this. Will fix and add in regression test.

**Comment by ntucker** [comment]
Curious which you think might be better designs to resolve #3

## Solutions to third problem:

### Custom applyMiddleware

This would take managers, and middleware. Managers would apply selector, middleware would not.

```typescript
import { createStore } from 'redux';
import { applyMiddleware } from 'rest-hooks';

const selector = state => state.restHooks;
const store = createStore(
  reducer,
  initialState,
  applyMiddleware(selector)(
    networkManager,
    subscriptionManager,
    // place Rest Hooks built middlewares before PromiseifyMiddleware
    PromiseifyMiddleware,
    // place redux middlewares after PromiseifyMiddleware
  ),
);
```

### Pass selector to getMiddleware

```typescript
import { createStore, applyMiddleware } from 'redux';

const selector = state => state.restHooks;
const store = createStore(
  reducer,
  initialState,
  applyMiddleware(
    networkManager.getMiddleware(selector),
    subscriptionManager.getMiddleware(selector),
    // place Rest Hooks built middlewares before PromiseifyMiddleware
    PromiseifyMiddleware,
    // place redux middlewares after PromiseifyMiddleware
  ),
);
```

**Comment by jaggaer-nkrunic** [comment]
Hey, thank you for your quick response and for looking into this.
As for design suggestions, i would discard #1 solution because it would break compatibility with redux-toolkit.
Speaking of which, i like their approach by having `getDefaultMiddleware(options)` function which returns default middleware and takes options where you can exclude middleware you don't want or pass extra argument to it, which is essentially your example #3 with some changes eg.

```js
import { configureStore } from '@reduxjs/toolkit';
import { createReduxMiddleware } from 'rest-hooks';

const selector = state => state.restHooks;
const store = configureStore({
  reducer: rootReducer,
  middleware: getDefaultMiddleware => {
    return [
      ...createReduxMiddleware({
        selector,
        networkManager: {
          defaultCacheExpiryLength: 1234,
        },
        subscriptionManager: {
          subscription: PollingSubscription,
        }
      }),
      ...getDefaultMiddleware({
        immutableCheck: false,
      })
    ];
  }
});
```
And implementation will be a bit more complex and would return already configured managers as middleware.
With that said i vote for #3 cause it can be extended later do something like this.
Does that make sense to you?

**Comment by ntucker** [comment]
@jaggaer-nkrunic Yes, this makes sense. Thanks for the insights. I'll go with the simpler version of #3 for now, and if you could create a feature request for the createReduxMiddleware() like getDefaultMiddleware() concept, we could eventually get around to providing that as well!

**Comment by ntucker** [comment]
@jaggaer-nkrunic Fixed released in [5.0.11](https://www.npmjs.com/package/rest-hooks/v/5.0.11), https://resthooks.io/docs/guides/redux docs updated.
